fossas/fossa-cli
# Ane 238 maven depgraph deep deps

# Overview

Maven projects weren't having top-level dependencies marked as a direct, instead they were all deep. The depgraph plugin, however, doesn't give any indication that a dep is deep except using the text formatted graph. This change switches the maven plugin strategy to use the text graph output.

## Acceptance criteria

1. `sourceUnits` should have direct dependencies in the top-level `Imports` field.
2. Project submodules should not be included and their immediate dependencies should be promoted to being direct.

## Testing plan

I used several projects to verify that this was working. The simplest was [clojure](https://github.com/clojure/clojure/archive/refs/tags/clojure-1.11.1.tar.gz) which has a tree of dependencies with no submodules. In addition, I used [zookeeper](https://www.apache.org/dyn/closer.lua/zookeeper/zookeeper-3.8.0/apache-zookeeper-3.8.0.tar.gz) and [log4j](https://www.apache.org/dyn/closer.lua/logging/log4j/2.17.2/apache-log4j-2.17.2-src.tar.gz).

To test:

1. Run fossa against one of those projects
2. Open `target/dependency-graph.txt` next to the output of fossa.
3. Check that the top-level `Imports` in `sourceUnits` are either at the top of the tree printed by the plugin or the direct dependencies of a submodule as show in the dependency graph. Submodules may not appear as a top-level item, in the case of submodules depending on eachother, they may be children of each other.
4. The list of submodules can be viewed by opening `pom.xml` and looking for the `<modules>` tag. These submodules should not appear in the output graph.

## Risks

This change will likely change the output for many existing customers because in the past we weren't marking any deps as direct. Is there anything else we can do to get assurance of its correctness?

## References

[ANE-238](https://fossa.atlassian.net/browse/ANE-238)

## Checklist
